// unlazy/dist/unlazy.with-hashing.js
const _t = "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7", ot = typeof window > "u", pt = !ot && "loading" in HTMLImageElement.prototype, Et = !ot && (!("onscroll" in window) || /(gle|ing|ro)bot|crawl|spider/i.test(navigator.userAgent));
function st(t, e = document) {
  return typeof t == "string" ? [...e.querySelectorAll(t)] : t instanceof Element ? [t] : [...t];
}
function bt(t, e) {
  let n, r;
  return t >= 1 ? (n = e, r = Math.round(e / t)) : (n = Math.round(e * t), r = e), { width: n, height: r };
}
function ht(t, e) {
  let n;
  return function(...r) {
    n && clearTimeout(n), n = setTimeout(() => {
      n = void 0, t(...r);
    }, e);
  };
}
function yt(t) {
  let { PI: e, min: n, max: r, cos: s, round: c } = Math, f = t[0] | t[1] << 8 | t[2] << 16, o = t[3] | t[4] << 8, l = (f & 63) / 63, p = (f >> 6 & 63) / 31.5 - 1, d = (f >> 12 & 63) / 31.5 - 1, u = (f >> 18 & 31) / 31, a = f >> 23, _ = (o >> 3 & 63) / 63, G = (o >> 9 & 63) / 63, w = o >> 15, P = r(3, w ? a ? 5 : 7 : o & 7), b = r(3, w ? o & 7 : a ? 5 : 7), C = a ? (t[5] & 15) / 15 : 1, I = (t[5] >> 4) / 15, H = a ? 6 : 5, g = 0, L = (U, y, D) => {
    let S = [];
    for (let z = 0; z < y; z++)
      for (let T = z ? 0 : 1; T * y < U * (y - z); T++)
        S.push(((t[H + (g >> 1)] >> ((g++ & 1) << 2) & 15) / 7.5 - 1) * D);
    return S;
  }, h = L(P, b, u), Q = L(3, 3, _ * 1.25), N = L(3, 3, G * 1.25), x = a && L(5, 5, I), B = Lt(t), M = c(B > 1 ? 32 : 32 * B), F = c(B > 1 ? 32 / B : 32), O = new Uint8Array(M * F * 4), q = [], Z = [];
  for (let U = 0, y = 0; U < F; U++)
    for (let D = 0; D < M; D++, y += 4) {
      let S = l, z = p, T = d, J = C;
      for (let i = 0, A = r(P, a ? 5 : 3); i < A; i++)
        q[i] = s(e / M * (D + 0.5) * i);
      for (let i = 0, A = r(b, a ? 5 : 3); i < A; i++)
        Z[i] = s(e / F * (U + 0.5) * i);
      for (let i = 0, A = 0; i < b; i++)
        for (let E = i ? 0 : 1, R = Z[i] * 2; E * b < P * (b - i); E++, A++)
          S += h[A] * q[E] * R;
      for (let i = 0, A = 0; i < 3; i++)
        for (let E = i ? 0 : 1, R = Z[i] * 2; E < 3 - i; E++, A++) {
          let et = q[E] * R;
          z += Q[A] * et, T += N[A] * et;
        }
      if (a)
        for (let i = 0, A = 0; i < 5; i++)
          for (let E = i ? 0 : 1, R = Z[i] * 2; E < 5 - i; E++, A++)
            J += x[A] * q[E] * R;
      let K = S - 2 / 3 * z, tt = (3 * S - K + T) / 2, At = tt - T;
      O[y] = r(0, 255 * n(1, tt)), O[y + 1] = r(0, 255 * n(1, At)), O[y + 2] = r(0, 255 * n(1, K)), O[y + 3] = r(0, 255 * n(1, J));
    }
  return { w: M, h: F, rgba: O };
}
function Lt(t) {
  let e = t[3], n = t[2] & 128, r = t[4] & 128, s = r ? n ? 5 : 7 : e & 7, c = r ? e & 7 : n ? 5 : 7;
  return s / c;
}
/**
 * Encodes an RGBA image to a PNG data URI. RGB should not be premultiplied by A.
 *
 * @remarks
 * This is optimized for speed and simplicity and does not optimize for size
 * at all. This doesn't do any compression (all values are stored uncompressed).
 *
 * @see https://github.com/evanw/thumbhash
 * @author Evan Wallace
 * @license MIT
 */
function ct(t, e, n) {
  const r = t * 4 + 1, s = 6 + e * (5 + r), c = [
    137,
    80,
    78,
    71,
    13,
    10,
    26,
    10,
    0,
    0,
    0,
    13,
    73,
    72,
    68,
    82,
    0,
    0,
    t >> 8,
    t & 255,
    0,
    0,
    e >> 8,
    e & 255,
    8,
    6,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    s >>> 24,
    s >> 16 & 255,
    s >> 8 & 255,
    s & 255,
    73,
    68,
    65,
    84,
    120,
    1
  ], f = [
    0,
    498536548,
    997073096,
    651767980,
    1994146192,
    1802195444,
    1303535960,
    1342533948,
    -306674912,
    -267414716,
    -690576408,
    -882789492,
    -1687895376,
    -2032938284,
    -1609899400,
    -1111625188
  ];
  let o = 1, l = 0;
  for (let d = 0, u = 0, a = r - 1; d < e; d++, a += r - 1)
    for (c.push(d + 1 < e ? 0 : 1, r & 255, r >> 8, ~r & 255, r >> 8 ^ 255, 0), l = (l + o) % 65521; u < a; u++) {
      const _ = n[u] & 255;
      c.push(_), o = (o + _) % 65521, l = (l + o) % 65521;
    }
  c.push(
    l >> 8,
    l & 255,
    o >> 8,
    o & 255,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    73,
    69,
    78,
    68,
    174,
    66,
    96,
    130
  );
  for (let [d, u] of [[12, 29], [37, 41 + s]]) {
    let a = -1;
    for (let _ = d; _ < u; _++)
      a ^= c[_], a = a >>> 4 ^ f[a & 15], a = a >>> 4 ^ f[a & 15];
    a = ~a, c[u++] = a >>> 24, c[u++] = a >> 16 & 255, c[u++] = a >> 8 & 255, c[u++] = a & 255;
  }
  return `data:image/png;base64,${globalThis.btoa(String.fromCharCode(...c))}`;
}
function wt(t) {
  const e = Ct(t), { w: n, h: r, rgba: s } = yt(e);
  return ct(n, r, s);
}
function Ct(t) {
  return Uint8Array.from(
    globalThis.atob(gt(t)),
    (e) => e.charCodeAt(0)
  );
}
function gt(t) {
  return t.replaceAll("-", "+").replaceAll("_", "/");
}
const xt = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz#$%*+,-.:;=?@[]^_{|}~", Y = (t, e, n) => {
  let r = 0;
  for (; e < n; )
    r *= 83, r += xt.indexOf(t[e++]);
  return r;
}, lt = Math.pow, v = Math.PI, zt = v * 2, at = 3294.6, it = 269.025, Pt = (t) => t > 10.31475 ? lt(t / it + 0.052132, 2.4) : t / at, V = (t) => ~~(t > 1227e-8 ? it * lt(t, 0.416666) - 13.025 : t * at + 1), m = (t) => (t < 0 ? -1 : 1) * t * t, rt = (t) => {
  for (t += v / 2; t > v; )
    t -= zt;
  const e = 1.27323954 * t - 0.405284735 * m(t);
  return 0.225 * (m(e) - e) + e;
};
function St(t) {
  const e = Y(t, 2, 6);
  return [e >> 16, e >> 8 & 255, e & 255];
}
function Tt(t, e, n, r) {
  const s = Y(t, 0, 1), c = s % 9 + 1, f = ~~(s / 9) + 1, o = c * f;
  let l = 0, p = 0, d = 0, u = 0, a = 0, _ = 0, G = 0, w = 0, P = 0, b = 0, C = 0, I = 0, H = 0, g = 0;
  const L = (Y(t, 1, 2) + 1) / 13446 * (r | 1), h = new Float64Array(o * 3), Q = St(t);
  for (l = 0; l < 3; l++)
    h[l] = Pt(Q[l]);
  for (l = 1; l < o; l++)
    g = Y(t, 4 + l * 2, 6 + l * 2), h[l * 3] = m(~~(g / (19 * 19)) - 9) * L, h[l * 3 + 1] = m(~~(g / 19) % 19 - 9) * L, h[l * 3 + 2] = m(g % 19 - 9) * L;
  const N = e * 4, x = new Uint8ClampedArray(N * n);
  for (u = 0; u < n; u++)
    for (I = v * u / n, d = 0; d < e; d++) {
      for (a = 0, _ = 0, G = 0, H = v * d / e, p = 0; p < f; p++)
        for (P = rt(I * p), l = 0; l < c; l++)
          w = rt(H * l) * P, b = (l + p * c) * 3, a += h[b] * w, _ += h[b + 1] * w, G += h[b + 2] * w;
      C = 4 * d + u * N, x[C] = V(a), x[C + 1] = V(_), x[C + 2] = V(G), x[C + 3] = 255;
    }
  return x;
}
function Gt(t, {
  ratio: e = 1,
  size: n = 32
} = {}) {
  const { width: r, height: s } = bt(e, n), c = Tt(t, r, s);
  return ct(r, s, c);
}
function ft(t = 'img[loading="lazy"]', {
  hash: e = !0,
  hashType: n = "blurhash",
  placeholderSize: r = 32,
  updateSizesOnResize: s = !1,
  onImageLoad: c
} = {}) {
  const f = /* @__PURE__ */ new Set();
  for (const o of st(t)) {
    const l = j(o, { updateOnResize: s });
    if (s && l && f.add(l), e) {
      const d = Ut({
        image: o,
        hash: typeof e == "string" ? e : void 0,
        hashType: n,
        size: r
      });
      d && (o.src = d);
    }
    if (!o.dataset.src && !o.dataset.srcset) {
      (typeof __UNLAZY_LOGGING__ > "u" || __UNLAZY_LOGGING__) && console.error("[unlazy] Missing `data-src` or `data-srcset` attribute", o);
      continue;
    }
    if (Et || !pt) {
      ut(o), k(o), $(o);
      continue;
    }
    if (o.src || (o.src = _t), o.complete && o.naturalWidth > 0) {
      X(o, c);
      continue;
    }
    const p = () => X(o, c);
    o.addEventListener("load", p, { once: !0 }), f.add(
      () => o.removeEventListener("load", p)
    );
  }
  return () => {
    for (const o of f)
      o();
    f.clear();
  };
}
function Ot(t = 'img[data-sizes="auto"], source[data-sizes="auto"]') {
  for (const e of st(t))
    j(e);
}
function X(t, e) {
  var o;
  if (((o = t.parentElement) == null ? void 0 : o.tagName.toLowerCase()) === "picture") {
    ut(t), k(t), $(t), e == null || e(t);
    return;
  }
  const r = new Image(), { srcset: s, src: c, sizes: f } = t.dataset;
  if (f === "auto") {
    const l = dt(t);
    l && (r.sizes = `${l}px`);
  } else
    t.sizes && (r.sizes = t.sizes);
  s && (r.srcset = s), c && (r.src = c), r.addEventListener("load", () => {
    k(t), $(t), e == null || e(t);
  }, { once: !0 });
}
function Ut({
  /** If given, the hash will be extracted from the image's `data-blurhash` or `data-thumbhash` attribute and ratio will be calculated from the image's actual dimensions */
  image: t,
  hash: e,
  hashType: n = "blurhash",
  /** @default 32 */
  size: r = 32,
  /** Will be calculated from the image's actual dimensions if not provided and image is given */
  ratio: s
} = {}) {
  if (!e && t) {
    const { blurhash: c, thumbhash: f } = t.dataset;
    e = f || c, n = f ? "thumbhash" : "blurhash";
  }
  if (e)
    try {
      if (n === "thumbhash")
        return wt(e);
      if (!s && t) {
        const c = t.width || t.offsetWidth || r, f = t.height || t.offsetHeight || r;
        s = c / f;
      }
      return Gt(e, { ratio: s, size: r });
    } catch (c) {
      (typeof __UNLAZY_LOGGING__ > "u" || __UNLAZY_LOGGING__) && console.error(`Error generating ${n} placeholder:`, c);
    }
}
const W = /* @__PURE__ */ new WeakMap();
function j(t, e) {
  var r;
  if (t.dataset.sizes !== "auto")
    return;
  const n = dt(t);
  if (n && (t.sizes = `${n}px`), ((r = t.parentElement) == null ? void 0 : r.tagName.toLowerCase()) === "picture" && !(e != null && e.skipChildren) && [...t.parentElement.getElementsByTagName("source")].forEach(
    (s) => j(s, { skipChildren: !0 })
  ), e != null && e.updateOnResize) {
    if (!W.has(t)) {
      const s = ht(() => j(t), 500), c = new ResizeObserver(s);
      W.set(t, c), c.observe(t);
    }
    return () => {
      const s = W.get(t);
      s && (s.disconnect(), W.delete(t));
    };
  }
}
function $(t) {
  t.dataset.src && (t.src = t.dataset.src, t.removeAttribute("data-src"));
}
function k(t) {
  t.dataset.srcset && (t.srcset = t.dataset.srcset, t.removeAttribute("data-srcset"));
}
function ut(t) {
  const e = t.parentElement;
  (e == null ? void 0 : e.tagName.toLowerCase()) === "picture" && ([...e.querySelectorAll("source[data-srcset]")].forEach(k), [...e.querySelectorAll("source[data-src]")].forEach($));
}
function dt(t) {
  var e, n;
  return t instanceof HTMLSourceElement ? (n = (e = t.parentElement) == null ? void 0 : e.getElementsByTagName("img")[0]) == null ? void 0 : n.offsetWidth : t.offsetWidth;
}
const Dt = Object.freeze({
  autoSizes: Ot,
  lazyLoad: ft,
  loadImage: X
});
var nt;
(nt = document.currentScript) != null && nt.hasAttribute("init") && ft();
export {
  Ot as autoSizes,
  Dt as default,
  ft as lazyLoad,
  X as loadImage
};
