// unlazy/dist/unlazy.js
const L = "data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7", p = typeof window > "u", b = !p && "loading" in HTMLImageElement.prototype, w = !p && (!("onscroll" in window) || /(gle|ing|ro)bot|crawl|spider/i.test(navigator.userAgent));
function z(e, t = document) {
  return typeof e == "string" ? [...t.querySelectorAll(e)] : e instanceof Element ? [e] : [...e];
}
function v(e, t) {
  let r;
  return function(...c) {
    r && clearTimeout(r), r = setTimeout(() => {
      r = void 0, e(...c);
    }, t);
  };
}
function S(e = 'img[loading="lazy"]', {
  hash: t = !0,
  hashType: r = "blurhash",
  placeholderSize: c = 32,
  updateSizesOnResize: n = !1,
  onImageLoad: i
} = {}) {
  const o = /* @__PURE__ */ new Set();
  for (const s of z(e)) {
    const a = f(s, { updateOnResize: n });
    if (n && a && o.add(a), !s.dataset.src && !s.dataset.srcset) {
      (typeof __UNLAZY_LOGGING__ > "u" || __UNLAZY_LOGGING__) && console.error("[unlazy] Missing `data-src` or `data-srcset` attribute", s);
      continue;
    }
    if (w || !b) {
      _(s), l(s), d(s);
      continue;
    }
    if (s.src || (s.src = L), s.complete && s.naturalWidth > 0) {
      A(s, i);
      continue;
    }
    const E = () => A(s, i);
    s.addEventListener("load", E, { once: !0 }), o.add(
      () => s.removeEventListener("load", E)
    );
  }
  return () => {
    for (const s of o)
      s();
    o.clear();
  };
}
function C(e = 'img[data-sizes="auto"], source[data-sizes="auto"]') {
  for (const t of z(e))
    f(t);
}
function A(e, t) {
  var s;
  if (((s = e.parentElement) == null ? void 0 : s.tagName.toLowerCase()) === "picture") {
    _(e), l(e), d(e), t == null || t(e);
    return;
  }
  const c = new Image(), { srcset: n, src: i, sizes: o } = e.dataset;
  if (o === "auto") {
    const a = y(e);
    a && (c.sizes = `${a}px`);
  } else
    e.sizes && (c.sizes = e.sizes);
  n && (c.srcset = n), i && (c.src = i), c.addEventListener("load", () => {
    l(e), d(e), t == null || t(e);
  }, { once: !0 });
}
const u = /* @__PURE__ */ new WeakMap();
function f(e, t) {
  var c;
  if (e.dataset.sizes !== "auto")
    return;
  const r = y(e);
  if (r && (e.sizes = `${r}px`), ((c = e.parentElement) == null ? void 0 : c.tagName.toLowerCase()) === "picture" && !(t != null && t.skipChildren) && [...e.parentElement.getElementsByTagName("source")].forEach(
    (n) => f(n, { skipChildren: !0 })
  ), t != null && t.updateOnResize) {
    if (!u.has(e)) {
      const n = v(() => f(e), 500), i = new ResizeObserver(n);
      u.set(e, i), i.observe(e);
    }
    return () => {
      const n = u.get(e);
      n && (n.disconnect(), u.delete(e));
    };
  }
}
function d(e) {
  e.dataset.src && (e.src = e.dataset.src, e.removeAttribute("data-src"));
}
function l(e) {
  e.dataset.srcset && (e.srcset = e.dataset.srcset, e.removeAttribute("data-srcset"));
}
function _(e) {
  const t = e.parentElement;
  (t == null ? void 0 : t.tagName.toLowerCase()) === "picture" && ([...t.querySelectorAll("source[data-srcset]")].forEach(l), [...t.querySelectorAll("source[data-src]")].forEach(d));
}
function y(e) {
  var t, r;
  return e instanceof HTMLSourceElement ? (r = (t = e.parentElement) == null ? void 0 : t.getElementsByTagName("img")[0]) == null ? void 0 : r.offsetWidth : e.offsetWidth;
}
const O = Object.freeze({
  autoSizes: C,
  lazyLoad: S,
  loadImage: A
});
var h;
(h = document.currentScript) != null && h.hasAttribute("init") && S();
export {
  C as autoSizes,
  O as default,
  S as lazyLoad,
  A as loadImage
};
